webapp/src/components/jira_epic_selector/jira_epic_selector.tsx

```diff
@@ -0,0 +1,239 @@
+        }
+        this.fetchInitialSelectedValues();
+    }
+
+    componentWillUnmount(): void {
+        if (this.props.removeValidate) {
+            this.props.removeValidate(this.isValid);
+        }
+    }
+
+    componentDidUpdate(prevProps: Props, prevState: State): void {
+        if (prevState.invalid && this.props.value !== prevProps.value) {
+            this.setState({invalid: false}); //eslint-disable-line react/no-did-update-set-state
+        }
+    }
+
+    fetchInitialSelectedValues = (): void => {
+        if (!this.props.value.length) {
+            return;
+        }
+
+        const epicIds = this.props.value.join(', ');
+        const searchStr = `and id IN (${epicIds})`;
+        const userInput = ''; // Fetching by saved ids, no user input to process
+
+        this.fetchEpicsFromJql(searchStr, userInput).then((options) => {
+            if (options) {
+                this.setState({cachedSelectedOptions: this.state.cachedSelectedOptions.concat(options)});
+            }
+        });
+    };
+
+    handleIssueSearchTermChange = (inputValue: string): Promise<ReactSelectOption[]> => {
+        return this.debouncedSearchIssues(inputValue);
+    };
+
+    searchIssues = async (userInput: string): Promise<ReactSelectOption[]> => {
+        const epicIssueType = this.props.issueMetadata.projects[0].issuetypes.find(isEpicIssueType);
+        if (!epicIssueType) {
+            return [];
+        }
+
+        const epicNameTypeId = Object.keys(epicIssueType.fields).find((key) => isEpicNameField(epicIssueType.fields[key]));
+        if (!epicNameTypeId) {
+            return [];
+        }
+
+        const epicNameTypeName = epicIssueType.fields[epicNameTypeId].name;
+
+        let searchStr = '';
+        if (userInput) {
+            const cleanedInput = userInput.trim().replace(/"/g, '\\"');
+            searchStr = ` and "${epicNameTypeName}"~"${cleanedInput}" `;
+        }
+
+        return this.fetchEpicsFromJql(searchStr, userInput);
+    };
+
+    debouncedSearchIssues = debounce(this.searchIssues, searchDebounceDelay);
+
+    fetchEpicsFromJql = async (jqlSearch: string, userInput: string): Promise<ReactSelectOption[]> => {
+        const epicIssueType = this.props.issueMetadata.projects[0].issuetypes.find(isEpicIssueType);
+        if (!epicIssueType) {
+            return [];
+        }
+
+        const epicNameTypeId = Object.keys(epicIssueType.fields).find((key) => isEpicNameField(epicIssueType.fields[key]));
+        if (!epicNameTypeId) {
+            return [];
+        }
+
+        const projectKey = this.props.issueMetadata.projects[0].key;
+        const fullJql = `project=${projectKey} and issuetype=${epicIssueType.id} ${jqlSearch} ${searchDefaults}`;
+
+        const params = {
+            jql: fullJql,
+            epic_name_type_id: epicNameTypeId,
+            q: userInput,
+        };
+
+        return this.props.fetchEpicsWithParams(params).then(({data}) => {
```

Comment:
Would it be worth it to pick one paradigm (async or promises) instead of mixing them? Eg, use try/catch here.  (this is a comment for all the code in this PR, basically)

Reply:
There's a pattern used in the actions and components in mattermost-webapp where `async/await` is used in the actions, and `then()` is used in components.

The Jira plugin has the same pattern. `async/await` is actually not used in any component in the repo. I think it's a good pattern to use `then()` in the components, because the callbacks  stand out more, and they are usually significant in handling success/error in the component's state.